### ISSUE TITLE
Switching to Audio Only?

### ISSUE THREAD

**Issue by mcorner** [twilio/video-quickstart-ios#299]
When the connection quality drops is there an easy way to switch to audio only?

We monitor videostats and we believe that on calls with poor connectivity it would probably be better to turn off video entirely and switch to audio only.  We are thinking about automating this ourselves, but we were hoping there was a solution already?

Unfortunately if we disable the video track there won't be any videostats so we wouldn't know when the connection improves.

This is p2p on the 2.x version across js, android, and ios.

Thanks!

**Comment by ceaglest** [comment]
Hi @mcorner,

Thanks for filing an issue. At the moment, we don't have any sort of automatic behavior to disable video Tracks when bandwidth becomes limited.

I have been looking into what is possible in Chromium WebRTC and this could be achieved by defining a custom bitrate allocation strategy which sets and enforces a minimum bitrate for `LocalVideoTrack`s. This is early on, but we might consider modifying our own `TVIEncodingParameters` to add new properties that support this use case. For example:

```.objc
@interface TVIEncodingParameters : NSObject

// Existing initializer, we will want to introduce a new one.
- (_Nonnull instancetype)initWithAudioBitrate:(NSUInteger)maxAudioBitrate videoBitrate:(NSUInteger)maxVideoBitrate;

// Existing API.
@property (nonatomic, assign, readonly) NSUInteger maxAudioBitrate;
@property (nonatomic, assign, readonly) NSUInteger maxVideoBitrate;

/**
 *  @brief The minimum video send bitrate in bits per second. Defaults to 0.
 *
 *  @discussion Audio always takes priority over video when allocating bandwidth. If a non-zero value is set,
 *  and there is not enough available bandwidth to satisfy the minimum bitrate requirement then `LocalVideoTrack`s
 *  will be disabled automatically.
 */
@property (nonatomic, assign, readonly) NSUInteger minVideoBitrate;

@end
```

If you wanted to achieve this today, I think your best bet would be to use `getStats:` in combination with `TVIIceCandidatePairStats`. You roughly want to:

1. Call getStats: periodically, perhaps 1/s to match `availableOutgoingBitrate`s computation window.
2. Find the active ICE candidate pair i.e. `TVIIceCandidatePairStats.isActiveCandidatePair == YES`.
3. Monitor `TVIIceCandidatePairStats.availableOutgoingBitrate`. If this drops below some threshold that you choose, and for long enough time then disable the LocalVideoTrack in response. Use a similar heuristic to re-enable if outgoing bandwidth returns.

Best,
Chris

**Comment by mcorner** [comment]
@ceaglest Thanks Chris.  We will need something that will work on Android as well and I am not seeing the equivalent ICE stats there?

**Comment by ceaglest** [comment]
Hey @mcorner,

The Android team is currently working on adding `IceCandidatePairStats` in an upcoming release to match the iOS SDK.

What did you think about the proposed API addition to `TVIEncodingParameters`. Would this satisfy your use case?

Best,
Chris

**Comment by mcorner** [comment]
@ceaglest Yes, the interface looks reasonable.  

Our current plan (until the Ice stats are in both platforms) is to try and use the total outgoing bandwidth from the stats callback to judge connection quality.  If it drops below 100kbps for N seconds we will disable video.  If it rises above 100kbps for M seconds we will reenable it.  Since audio will consume more than 100kbps on its own when the connection is good, this seems like it might work, but we will have to play around with it and get the tuning right.

In your interface/implementation and ours we  have to get the hysteresis right or it will bounce between video and no video which won't be a great experience.

**Comment by treyrich** [comment]
I also think that the proposed interface would solve the problem quite well. Is there an ETA on that being rolled out?

**Comment by mcorner** [comment]
Just to update, our plan didn't really work.  If the connection gets cut down to 30kbs we disable the video.  But then if the connection improves (say to 1MBps) the audio doesn't seem to use any more than the 30kbps.  So we ended up implementing a button to retry enabling video.  This won't work for many use cases of course.

**Comment by treyrich** [comment]
We ran into the same problem. It's fairly easy to disable the video on a poor connection, but we've yet to find a consistent way to reenable it on network recovery.

**Comment by ceaglest** [comment]
Hey @mcorner and @treyrich,

This is why it would be better to use estimates on available bandwidth rather than relying on the bandwidth consumed in audio. There are limits to how much B/W is consumed depending on the codec which you choose (OPUS, iSAC etc.). Unfortunately, I did go ahead and check `TVIIceCandidatePairStats.availableOutgoingBitrate`, and we do not fill out this value. It's part of the W3C spec, but is not produced by our implementation at the moment.

I do have a prototype which integrates directly with PeerConnection, rather than consuming stats. This is showing some promise but there is still more work to do. We are currently considering how to handle Track switch-off in both Group and Peer-to-Peer Rooms, and the API implications outside of `TVIEncodingParameters`.

I'll share more info as I have it.

Thanks,
Chris

**Comment by mcorner** [comment]
@ceaglest I thought I would check in on this one.  

I noticed the network quality announcement: https://www.twilio.com/docs/video/using-network-quality-api . But....we are using p2p because of handoff problems we had between wifi and cellular in group rooms (has that changed?), and network quality isn't supported on Android and iOS.  

Are the underlying availableOutgoingBandwidth available on iOS and/or Android yet?

**Comment by ceaglest** [comment]
Hi @mcorner,

We are currently working on Network Quality APIs for mobile. As you noticed, this feature is enabled by our Media Servers in Group Rooms. We don't currently have plans to support Network Quality in Peer-to-Peer Rooms, though this could change based upon demand.

> But....we are using p2p because of handoff problems we had between wifi and cellular in group rooms (has that changed?)

There is more info in older tickets, but initially our Media Servers did not support ICE restart, which prevented network handoff (Wifi <-> Cell) from working. This was fixed, and we also identified and made several improvements to our Client's behavior in the 2.5.x releases. If you're seeing a situation where handoff is not working in Peer-to-Peer or Group Rooms we would definitely like to know about it. I would be happy to look at any Room SID(s) that can provide.

> Are the underlying availableOutgoingBandwidth available on iOS and/or Android yet?

I'm sorry to say they are not yet. I will look into this and get back to you.

Thanks,
Chris

**Comment by mcorner** [comment]
Ok, thanks for the update. Once network quality is available on mobile, we will give group rooms a retry.

**Comment by ceaglest** [comment]
Hi @mcorner,

I have some updates from the past year that are relevant to this discussion.

1. Network quality for local and remote Participants is now available on iOS, and received a bug fix in [3.2.1](https://www.twilio.com/docs/video/changelog-twilio-video-ios-version-3x#321-february-10-2020). Android gained full support as of [5.2.0](https://www.twilio.com/docs/video/changelog-twilio-video-android#520-march-9th-2020). This requires using a Group Room.
2. `TVIIceCandidatePairStats.availableOutgoingBitrate` is now correctly populated as of iOS [3.2.0](https://www.twilio.com/docs/video/changelog-twilio-video-ios-version-3x#320-february-5-2020) / Android [5.1.1](https://www.twilio.com/docs/video/changelog-twilio-video-android#511-february-14th-2020). You can use this property in Peer-to-Peer (or Group) Rooms to determine when there is insufficient bandwidth for video, and disable/unpublish video in response.

Hopefully this will allow you to provide a better experience, instead of having to use a retry button.

We don't have plans to offer automatic Track switch-off for publishers in the near future. However, are in the process of adding [Bandwidth Profile APIs](https://www.twilio.com/docs/video/tutorials/developing-high-quality-video-applications) (to match the JS SDK) which allows for subscriber side Track switch-off in Group Rooms.

Let me know if you need anything else for disabling publisher video.

Thanks,
Chris

**Comment by ceaglest** [comment]
Closing, but fell free to reopen if you any more related questions about stats or network quality APIs.
